torch/csrc/jit/codegen/cuda/lower_insert_syncs.cpp

```diff
@@ -68,244 +72,255 @@ class SmemAllocMap {
+    auto fl_i = std::distance(for_loops_.begin(), fl_it) + 1;
 
-  static bool detectIntersection(const TvSet& left, const TvSet& right) {
-    for (auto item : left) {
-      if (right.find(item) != right.end()) {
+    // Start at that index and see if there's syncs within that for loop
+    for (auto i : c10::irange(fl_i, sync_hit_.size())) {
+      if (sync_hit_[i]) {
         return true;
       }
     }
     return false;
   }
 
-  void addOutputSmemTvs(const kir::Expr* expr, TvSet& set) {
-    for (auto out : expr->outputs()) {
-      if (auto tv = dynamic_cast<kir::TensorView*>(out)) {
-        if (tv->memoryType() == MemoryType::Shared) {
-          auto real_tv = alloc_map_.getRealBuffer(tv);
-          set.insert(real_tv);
-        }
+  void handle(kir::Expr* expr) final {
+    // If not a tensor view expression continue with dispatch
+    if (!ir_utils::isTVOp(expr)) {
+      kir::ExprMutator::handle(expr);
+      return;
+    }
+
+    // Mark write has been hit for all output tvs
+    auto out_tvs = ir_utils::filterByType<kir::TensorView>(expr->outputs());
+    for (auto out_tv : out_tvs) {
+      if (out_tv->memoryType() != MemoryType::Shared) {
+        continue;
       }
+      auto& entry = getMemInfo(out_tv);
+
+      // If this is the first write and there's a sync in one of the loops after
+      // the compute at loop, then this buffer is protected.
+      if (syncWithin(entry.ca_loop) && !entry.write_hit) {
+        entry.sync_before_write = true;
+      }
+      entry.write_hit = true;
     }
-  }
 
-  void addInputSmemTvs(const kir::Expr* expr, TvSet& set) {
-    for (auto in : expr->inputs()) {
-      if (auto tv = dynamic_cast<kir::TensorView*>(in)) {
-        if (tv->memoryType() == MemoryType::Shared) {
-          auto real_tv = alloc_map_.getRealBuffer(tv);
-          set.insert(real_tv);
-        }
+    // Mark read was hit, if sync_after_read was set, clear it.
+    auto inp_tvs = ir_utils::filterByType<kir::TensorView>(expr->inputs());
+    for (auto inp_tv : inp_tvs) {
+      if (inp_tv->memoryType() != MemoryType::Shared) {
+        continue;
       }
+      auto& entry = getMemInfo(inp_tv);
+      entry.read_hit = true;
+      // Clear the sync_after_read if it was set because there was another write
+      entry.sync_after_read = false;
     }
   }
 
- private:
-  // Track which for loop was passed to the constructor to prevent recursive
-  // entrance. WAR for how this pass is structured.
-  const kir::ForLoop* base_fl_;
-
-  //! Allocation map of SMEM buffers
-  SmemAllocMap& alloc_map_;
+  void handle(kir::ForLoop* for_loop) final {
+    // Push loop scope information
+    auto prev_within_iter_loop_ = within_iter_loop_;
+    sync_hit_.push_back(false);
 
-  //! Track Shared Memory Inputs (Reads) for parent for-loop
-  TvSet all_smem_inputs_;
+    // If there is no real iterating loop WAR syncs aren't necessary
+    within_iter_loop_ = within_iter_loop_ ||
```

Comment:
Is sync necessary when it's `isThreadDim()`?

One of the differences with `devel` happens with `SmemIndexingSimple`. 

`devel`:
```
__global__ void kernel423(Tensor<float, 2> T0, Tensor<float, 2> T3, Tensor<float, 2> T2) {
  alignas(4) extern __shared__ char array[];
  unsigned offset = 0;
  offset = alignBufferSize(offset,4);
  float* T1 = reinterpret_cast<float*>(array + offset);
  offset += (T0.size[1] * sizeof(float));
  // Allocate global tensor T2
  T1[((nvfuser_index_t)threadIdx.x)]
    = T0[(((nvfuser_index_t)blockIdx.x) * T0.stride[0]) + (((nvfuser_index_t)threadIdx.x) * T0.stride[1])]
    + (float) 1;
  __barrier_sync(0);
  T2[(((nvfuser_index_t)blockIdx.x) * (1 * T0.size[1])) + (((nvfuser_index_t)threadIdx.x) * 1)]
    = T1[((nvfuser_index_t)threadIdx.x)]
    + (float) 1;
  T3[(((nvfuser_index_t)blockIdx.x) * (1 * T0.size[1])) + (((nvfuser_index_t)threadIdx.x) * 1)]
    = T2[(((nvfuser_index_t)blockIdx.x) * (1 * T0.size[1])) + (((nvfuser_index_t)threadIdx.x) * 1)]
    + (float) 1;
}
```

New:
```

__global__ void kernel423(Tensor<float, 2> T0, Tensor<float, 2> T3, Tensor<float, 2> T2) {
  alignas(4) extern __shared__ char array[];
  unsigned offset = 0;
  offset = alignBufferSize(offset,4);
  float* T1 = reinterpret_cast<float*>(array + offset);
  offset += (T0.size[1] * sizeof(float));
  // Allocate global tensor T2
  T1[((nvfuser_index_t)threadIdx.x)]
    = T0[(((nvfuser_index_t)blockIdx.x) * T0.stride[0]) + (((nvfuser_index_t)threadIdx.x) * T0.stride[1])]
    + (float) 1;
  __barrier_sync(0);
  T2[(((nvfuser_index_t)blockIdx.x) * (1 * T0.size[1])) + (((nvfuser_index_t)threadIdx.x) * 1)]
    = T1[((nvfuser_index_t)threadIdx.x)]
    + (float) 1;
  T3[(((nvfuser_index_t)blockIdx.x) * (1 * T0.size[1])) + (((nvfuser_index_t)threadIdx.x) * 1)]
    = T2[(((nvfuser_index_t)blockIdx.x) * (1 * T0.size[1])) + (((nvfuser_index_t)threadIdx.x) * 1)]
    + (float) 1;
  __barrier_sync(0);
}
```

The last barrier is not necessary, but I think that's added because `within_iter_loop_` is true.

Reply:
Yeah, I think it should just be isThread not isBlockDim.

Reply:
Any cases that are because of "isThread" should be RAW syncs.

Reply:
Pushed the fix.